File: neuralnetwork.js
const NeuralNetwork = (function() {

    class NeuralNetwork {
        constructor(inputNodes, layers, learningRate) {
            this.inputNodes = inputNodes || 1;
            this.layers = layers;
            this.learningRate = learningRate || 0.1;

            for(let i = 0; i < this.layers.length; i++) {
                this.layers[i].weights = new Matrix(this.layers[i].nodes, i === 0 ? this.inputNodes : this.layers[i - 1].nodes);
                this.layers[i].bias = new Matrix(this.layers[i].nodes, 1);

                this.layers[i].weights.random(-1, 1);
                this.layers[i].bias.random(-1, 1);
            }
        }

        predict(inputs) {
            inputs = Matrix.from_arr(inputs);

            let prevLayerValues = inputs;

            for(let layer of this.layers) {
                layer.layerValues = Matrix.multiply(layer.weights, prevLayerValues);
                layer.layerValues.add(layer.bias);
                layer.layerValues.map(layer.activationFunction.func);

                prevLayerValues = layer.layerValues;
            }

            const outputs = prevLayerValues;
            return Matrix.to_arr(outputs);
        }

        train(inputArray, targetArray) {
            const inputs = Matrix.from_arr(inputArray);

            let prevLayerValues = inputs;

            for(let layer of this.layers) {
                layer.layerValues = Matrix.multiply(layer.weights, prevLayerValues);
                layer.layerValues.add(layer.bias);
                layer.layerValues.map(layer.activationFunction.func);

                prevLayerValues = layer.layerValues;
            }

            const outputs = this.layers[this.layers.length - 1].layerValues;
            const targets = Matrix.from_arr(targetArray);

            let prevError = null;

            for(let i = this.layers.length - 1; i >= 0; i--) {
                const layer = this.layers[i];

                let error;

                if(!prevError) {
                    error = Matrix.substract(targets, outputs);
                } else {
                    const nextWeights = this.layers[i + 1].weights;
                    const nextWeightsTransposed = Matrix.transpose(nextWeights);

                    error = Matrix.multiply(nextWeightsTransposed, prevError);
                }

                const gradient = Matrix.map(layer.layerValues, layer.activationFunction.dfunc);
                gradient.multiply(error);
                gradient.multiply(this.learningRate);

                const prevValues = i === 0 ? inputs : this.layers[i - 1].layerValues;
                const prevValuesTransposed = Matrix.transpose(prevValues);
                const layerWeightsDeltas = Matrix.multiply(gradient, prevValuesTransposed);

                layer.weights.add(layerWeightsDeltas);
                layer.bias.add(gradient);

                prevError = error;
            }
        }

        setLearningRate(learningRate) {
            this.learningRate = learningRate;
        }

        serialize() {
            return JSON.stringify(this);
        }

        static deserialize(data) {
            if(typeof data === 'string') {
                data = JSON.parse(data);
            }
            let layers = [];

            for(let layer of data.layers) {
                layers[i] = Layer.deserialize(layer);
            }

            const neuralnetwork = NeuralNetwork.createNeuralNetwork({
                inputNodes: data.inputNodes,
                layers: layers,
                learningRate: data.learningRate
            });

            return neuralnetwork;
        }
    }

    class Layer {
        constructor(nodes, activationFunction) {
            this.nodes = nodes;
            this.activationFunction = activationFunction;
            this.weights = null;
            this.bias = null;
            this.layerValues = null;
        }

        static deserialize(data) {
            if(typeof data === 'string') {
                data = JSON.parse(data);
            }

            const layer = NeuralNetwork.createLayer({
                nodes: data.nodes,
                activation: data.activation
            });
            layer.weights = data.weights;
            layer.bias = data.bias;
            layer.layerValues = layerValues;

            return layer;
        }
    }

    function createNeuralNetwork(config) {
        return new NeuralNetwork(config.inputNodes, config.layers, config.learningRate);
    }

    function createLayer(config) {
        const activation = {};

        switch(config.activation) {
            case 'sigmoid':
                activation.func = sigmoid;
                activation.dfunc = dsigmoid;
            break;
            default:
                activation.func = sigmoid;
                activation.dfunc = dsigmoid;
        }

        return new Layer(config.nodes, activation);
    }

    function sigmoid(x) {
        return 1 / (1 + Math.exp(-x));
    }
    
    function dsigmoid(y) {
        return y * (1 - y);
    }

    return {
        createNeuralNetwork,
        createLayer
    };

})();

class Matrix {
    constructor(rows, cols) {
        this.rows = rows;
        this.cols = cols;

        this.matrix = [];

        for(let i = 0; i < this.rows; i++) {
            this.matrix[i] = [];
            for(let j = 0; j < this.cols; j++) {
                this.matrix[i][j] = 0;
            }
        }
    }

    random(min, max) {
        for(let i = 0; i < this.rows; i++) {
            for(let j = 0; j < this.cols; j++) {
                this.matrix[i][j] = Math.random() * (max - min) + min;
            }
        }
    }

    print() {
        console.table(this.matrix);
    }

    map(f) {
        for(let i = 0; i < this.rows; i++) {
            for(let j = 0; j < this.cols; j++) {
                let val = this.matrix[i][j];
                this.matrix[i][j] = f(val, i, j);
            }
        }
    }

    add(n) {
        if(n instanceof Matrix) {
            if(this.rows != n.rows || this.cols != n.cols) {
                throw 'Matrices are not the same size';
            }

            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] += n.matrix[i][j];
                }
            }
        } else {
            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] += n;
                }
            }
        } 
    }

    substract(n) {
        if(n instanceof Matrix) {
            if(this.rows != n.rows || this.cols != n.cols) {
                throw 'Matrices are not the same size';
            }

            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] -= n.matrix[i][j];
                }
            }
        } else {
            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] -= n;
                }
            }
        } 
    }

    multiply(n) {
        if(n instanceof Matrix) {  
            if(this.rows !== n.rows || this.cols !== n.cols) {
                throw "these matrices have to be the same size";
            }

            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] *= n.matrix[i][j];
                }
            }
        } else {

            for(let i = 0; i < this.rows; i++) {
                for(let j = 0; j < this.cols; j++) {
                    this.matrix[i][j] *= n;
                }
            }
        }
    }

    static transpose(m) {
        let result = new Matrix(m.cols, m.rows);

        for(let i = 0; i < result.rows; i++) {
            for(let j = 0; j < result.cols; j++) {
                result.matrix[i][j] = m.matrix[j][i];
            }
        }

        return result;
    }

    static add(m, n) {
        let result = new Matrix(m.rows, m.cols);

        if(n instanceof Matrix) {
            if(m.rows !== n.rows || m.cols !== n.cols) {
                throw 'Matrices are not the same size';
            }

            for(let i = 0; i < result.rows; i++) {
                for(let j = 0; j < result.cols; j++) {
                    result.matrix[i][j] = m.matrix[i][j] + n.matrix[i][j];
                }
            }

            return result;
        } else {
            for(let i = 0; i < result.rows; i++) {
                for(let j = 0; j < result.cols; j++) {
                    result.matrix[i][j] = m.matrix[i][j] + n;
                }
            }
        } 

        return result;
    }

    static substract(m, n) {
        let result = new Matrix(m.rows, m.cols);

        if(n instanceof Matrix) {
            if(m.rows !== n.rows || m.cols !== n.cols) {
                throw 'Matrices are not the same size';
            }

            for(let i = 0; i < result.rows; i++) {
                for(let j = 0; j < result.cols; j++) {
                    result.matrix[i][j] = m.matrix[i][j] - n.matrix[i][j];
                }
            }

            return result;
        } else {
            for(let i = 0; i < result.rows; i++) {
                for(let j = 0; j < result.cols; j++) {
                    result.matrix[i][j] = m.matrix[i][j] - n;
                }
            }
        } 

        return result;
    } 

    static multiply(a, b) {
        if(!isNaN(a) && !isNaN(b)) {
            return a * b;
        }

        if(a && b instanceof Matrix) {
            if(a.cols === b.rows) {
                let result = new Matrix(a.rows, b.cols);

                for(let i = 0; i < result.rows; i++) {
                    for(let j = 0; j < result.cols; j++) {
                        let sum = 0;
                        for(let k = 0; k < a.cols; k++) {
                            sum += a.matrix[i][k] * b.matrix[k][j];
                        }
                        result.matrix[i][j] = sum;
                    }
                }

                return result;

            } else if(a.rows === b.rows && a.cols === b.cols) {
                let result = new Matrix(a.rows, a.cols);

                for(let i = 0; i < result.rows; i++) {
                    for(let j = 0; j < result.cols; j++) {
                        result.matrix[i][j] = a.matrix[i][j] * b.matrix[i][j];
                    }
                }

                return result;
            } else {
                throw 'These matrices cannot be multiplied';
            }

        } else {
            let m = a;
            let n = b;

            if(!isNaN(a)) {
                m = b;
                n = a;
            }

            let result = new Matrix(m.rows, m.cols);

            for(let i = 0; i < m.rows; i++) {
                for(let j = 0; j < m.cols; j++) {
                    result.matrix[i][j] = m.matrix[i][j] * n;
                }
            }

            return result;
        }
    }

    static from_arr(arr, cols) {
        if(!cols) {
            cols = 1;
        }
        let rows = Math.ceil(arr.length / cols);

        let result = new Matrix(rows, cols);

        for(let i = 0; i < rows; i++) {
            for(let j = 0; j < cols; j++) {
                let index = i * cols + j;
                if(isNaN(arr[index])) {
                    result.matrix[i][j] = 0;
                } else {
                    result.matrix[i][j] = arr[index];
                }
            }
        }

        return result;
    }

    static to_arr(m) {
        let arr = [];

        for(let i = 0; i < m.rows; i++) {
            for(let j = 0; j < m.cols; j++) {
                arr.push(m.matrix[i][j]);
            }
        }

        return arr;
    }

    static map(m, f) {
        let result = new Matrix(m.rows, m.cols);
        result.map((e, i, j) => f(m.matrix[i][j]));

        return result;
    }

    serialize() {
        return JSON.stringify(this);
    }

    static deserialize(data) {
        if(typeof data === 'string') {
            data = JSON.parse(data);
        }

        let m = new Matrix(data.rows, data.cols);
        m.matrix = data.matrix;

        return m;
    }
}

File: python/neuralnetwork.py
import random
import math
from numbers import Number

class NeuralNetwork:
    def __init__(self, input_nodes, layers, learning_rate = 0.01):
        self.input_nodes = input_nodes
        self.layers = layers
        self.learning_rate = learning_rate

        for x in range(len(self.layers)):
            if x == 0:
                nodes = self.input_nodes
            else:
                nodes = self.layers[x - 1].nodes

            self.layers[x].weights = Matrix(self.layers[x].nodes, nodes)
            self.layers[x].bias = Matrix(self.layers[x].nodes, 1)

            self.layers[x].weights.random(0, 1)
            self.layers[x].bias.random(0, 1)

    def predict(self, input_arr):
        inputs = Matrix.from_arr(input_arr)

        prev_layer_values = inputs

        for layer in self.layers:
            layer.layer_values = Matrix._multiply(layer.weights, prev_layer_values)
            layer.layer_values.add(layer.bias)
            layer.layer_values.map(layer.activation_function.function)

            prev_layer_values = layer.layer_values

        outputs = prev_layer_values
        
        return Matrix.to_arr(outputs)

    def train(self, input_arr, target_arr):
        inputs = Matrix.from_arr(input_arr)

        prev_layer_values = inputs

        for layer in self.layers:
            layer.layer_values = Matrix._multiply(layer.weights, prev_layer_values)
            layer.layer_values.add(layer.bias)
            layer.layer_values.map(layer.activation_function.function)

            prev_layer_values = layer.layer_values

        outputs = prev_layer_values
        targets = Matrix.from_arr(target_arr)

        prev_error = None

        for x in range(len(self.layers)):
            index = len(self.layers) - 1 - x

            layer = self.layers[index]

            if prev_error == None:
                error = Matrix._subtract(targets, outputs)
            else:
                next_weights = self.layers[index + 1].weights
                next_weights_transposed = Matrix.transpose(next_weights)

                error = Matrix._multiply(next_weights_transposed, prev_error)
        
            gradient = Matrix._map(layer.layer_values, layer.activation_function.dfunction)
            gradient.multiply(error)
            gradient.multiply(self.learning_rate)

            prev_values = self.layers[index - 1].layer_values
            if index == 0:
                prev_values = inputs
            
            prev_values_transposed = Matrix.transpose(prev_values)
            layer_weights_deltas = Matrix._multiply(gradient, prev_values_transposed)

            layer.weights.add(layer_weights_deltas)
            layer.bias.add(gradient)

            prev_error = error

    def set_learning_rate(self, learning_rate):
        self.learning_rate = learning_rate


class Layer:
    def __init__(self, nodes, activation_function):
        self.nodes = nodes
        self.weights = None
        self.bias = None
        self.layer_values = None

        if activation_function == 'sigmoid':
            self.activation_function = ActivationFunction(sigmoid, dsigmoid)
        else:
            self.activation_function = ActivationFunction(sigmoid, dsigmoid)

class ActivationFunction:
    def __init__(self, function, dfunction):
        self.function = function
        self.dfunction = dfunction
        
def sigmoid(x):
    return 1 / (1 + math.exp(-x))

def dsigmoid(y):
    return y * (1 - y)

class Matrix:
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols

        self.matrix = []

        for x in range(self.rows):
            self.matrix.append([])
            for y in range(self.cols):
                self.matrix[x].append(0)
    
    def random(self, _min, _max):
        for x in range(self.rows):
            for y in range(self.cols):
                self.matrix[x][y] = random.random() * (_max - _min) + _min
    
    def map(self, f):
        for x in range(self.rows):
            for y in range(self.cols):
                value = self.matrix[x][y]
                self.matrix[x][y] = f(value)

    def add(self, n):
        if isinstance(n, Matrix):
            if self.rows != n.rows or self.cols != n.cols:
                raise Exception("Given matrices are not the same size")

            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] + n.matrix[x][y]
        
        else:
            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] + n

    def subtract(self, n):
        if isinstance(n, Matrix):
            if self.rows != n.rows or self.cols != n.cols:
                raise Exception("Given matrices are not the same size")

            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] - n.matrix[x][y]
        
        else:
            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] - n
    
    def multiply(self, n):
        if isinstance(n, Matrix):
            if self.rows != n.rows or self.cols != n.cols:
                raise Exception("Given matrices are not the same size")

            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] * n.matrix[x][y]
        
        else:
            for x in range(self.rows):
                for y in range(self.cols):
                    self.matrix[x][y] = self.matrix[x][y] * n

    @staticmethod
    def transpose(m):
        result = Matrix(m.cols, m.rows)

        for x in range(result.rows):
            for y in range(result.cols):
                result.matrix[x][y] = m.matrix[y][x]

        return result

    @staticmethod
    def _add(m, n):
        result = Matrix(m.rows, m.cols)

        if isinstance(n, Matrix):
            if m.rows != n.rows or m.cols != n.cols:
                raise Exception("Given matrices are not the same size")

            for x in range(result.rows):
                for y in range(result.cols):
                    result.matrix[x][y] = m.matrix[x][y] + n.matrix[x][y]

        else:
            for x in range(result.rows):
                for y in range(result.cols):
                    result.matrix[x][y] = m.matrix[x][y] + n

        return result

    @staticmethod
    def _subtract(m, n):
        result = Matrix(m.rows, m.cols)

        if isinstance(n, Matrix):
            if m.rows != n.rows or m.cols != n.cols:
                raise Exception("Given matrices are not the same size")

            for x in range(result.rows):
                for y in range(result.cols):
                    result.matrix[x][y] = m.matrix[x][y] - n.matrix[x][y]

        else:
            for x in range(result.rows):
                for y in range(result.cols):
                    result.matrix[x][y] = m.matrix[x][y] - n

        return result

    @staticmethod
    def _multiply(a, b):
        if isinstance(a, Number) and isinstance(b, Number):
            return a * b
        
        if isinstance(a, Matrix) and isinstance(b, Matrix):
            if a.cols == b.rows:
                result = Matrix(a.rows, b.cols)

                for x in range(result.rows):
                    for y in range(result.cols):
                        _sum = 0
                        for k in range(a.cols):
                            _sum += a.matrix[x][k] * b.matrix[k][y]

                        result.matrix[x][y] = _sum

                return result
            
            elif a.rows == b.rows and a.cols == b.cols:
                result = Matrix(a.rows, a.cols)

                for x in range(a.rows):
                    for y in range(a.cols):
                        result.matrix[x][y] = a.matrix[x][y] * b.matrix[x][y]

                return result
            
            else:
                raise Exception("Given matrices cannot be multiplied")
        
        else:
            m = a
            n = b

            if isinstance(a, Number):
                m = b
                n = a
            
            result = Matrix(m.rows, m.cols)

            for x in range(m.rows):
                for y in range(m.cols):
                    result.matrix[x][y] = m.matrix[x][y] * n

            return result

    @staticmethod
    def _map(m, f):
        result = Matrix(m.rows, m.cols)

        for x in range(m.rows):
            for y in range(m.cols):
                value = m.matrix[x][y]
                result.matrix[x][y] = f(value)

        return result

    @staticmethod
    def from_arr(arr, cols = 1):
        rows = math.ceil(len(arr) / cols)

        result = Matrix(rows, cols)

        for x in range(rows):
            for y in range(cols):
                index = x * cols + y
                if not isinstance(arr[index], Number):
                    result.matrix[x][y] = 0
                else:
                    result.matrix[x][y] = arr[index]

        return result

    @staticmethod
    def to_arr(m):
        arr = []

        for x in range(m.rows):
            for y in range(m.cols):
                arr.append(m.matrix[x][y])

        return arr